common/vsock.c

```diff
@@ -0,0 +1,119 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above
+ *       copyright notice, this list of conditions and the
+ *       following disclaimer.
+ *     * Redistributions in binary form must reproduce the
+ *       above copyright notice, this list of conditions and
+ *       the following disclaimer in the documentation and/or
+ *       other materials provided with the distribution.
+ *     * The names of contributors to this software may not be
+ *       used to endorse or promote products derived from this
+ *       software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * Author: David Woodhouse <dwmw2@infradead.org>
+ */
+
+#include "config.h"
+
+#include "vsock.h"
+
+#include <limits.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#ifdef HAVE_VSOCK
+#include <sys/socket.h>
+#include <linux/vm_sockets.h>
+#include <sys/ioctl.h>
+#endif
+
+/* This generic parsing utility doesn't actually require the
+ * vm_sockets.h header and thus doesn't require conditional
+ * compiliation... except for this one definition. */
+#ifndef VMADDR_CID_ANY
+#define VMADDR_CID_ANY -1U
+#endif
+
+bool
+p11_vsock_parse_addr (const char *target,
+		      unsigned int *cid,
+		      unsigned int *port)
+{
+	bool cid_found = false;
+	bool port_found = false;
+	unsigned long val;
+	char *endptr;
+
+	/* If the caller already set the CID to 'any' then
+	 * don't require it in the given address. */
+	if (*cid == VMADDR_CID_ANY)
+		cid_found = true;
+
+	while (*target) {
+		if (strncmp(target, "cid=", 4) == 0) {
+			val = strtoul(target + 4, &endptr, 0);
+			if (val > UINT_MAX || endptr == target + 4)
+				return false;
+			*cid = val;
+			cid_found = true;
+		} else if (strncmp(target, "port=", 5) == 0) {
+			val = strtoul(target + 5, &endptr, 0);
+			if (val > UINT_MAX || endptr == target + 5)
+				return false;
+			*port = val;
+			port_found = true;
+		} else {
+			return false;
+		}
+
+		target = endptr;
+		if (*target == ';')
+			target++;
+		else if (*target)
```

Comment:
Is this `else` clause needed?

Reply:
Yes, because it catches the `vsock:cid=500port=500` case without the semicolon.

The acceptable things to follow a valid parameter are either a semicolon, or the end of the string.